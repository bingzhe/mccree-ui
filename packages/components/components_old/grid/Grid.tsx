import * as React from "react";
import classNames from "classnames";
import { ConfigContext } from "../config-provider";
import { GridContext } from "./GridContext";
import { Breakpoint } from "./responsiveObserve";
import GridContainer from "./GridContainer";
import { GridFC, GridSize, FlexType } from "./Grid.type";

const { useContext } = React;

const parseFlex = (flex: FlexType): string => {
    if (typeof flex === "number") {
        return `${flex} ${flex} auto`;
    }

    if (/^\d+(\.\d+)?(px|em|rem|%)$/.test(flex)) {
        return `0 0 ${flex}`;
    }

    return flex;
};

const Grid: GridFC = (props) => {
    const {
        span,
        order,
        offset,
        push,
        pull,
        className,
        children,
        flex,
        style: styleProp,
        ...restProps
    } = props;

    const { getPrefixCls } = useContext(ConfigContext);
    const { gutter } = useContext(GridContext);
    const prefixCls = getPrefixCls("grid");

    let sizeClassObj = {};
    let style: React.CSSProperties = { ...styleProp };

    ["xs", "sm", "md", "lg", "xl", "xxl"].forEach((size: Breakpoint) => {
        let sizeProps: GridSize = {};
        const propSize = props[size];

        if (typeof propSize === "number") {
            sizeProps.span = propSize;
        } else if (typeof propSize === "object") {
            sizeProps = propSize || {};
        }

        delete restProps[size];

        sizeClassObj = {
            ...sizeClassObj,
            [`${prefixCls}-${size}-${sizeProps.span}`]: sizeProps.span !== undefined,
            [`${prefixCls}-${size}-order-${sizeProps.order}`]:
                sizeProps.order || sizeProps.order === 0,
            [`${prefixCls}-${size}-offset-${sizeProps.offset}`]:
                sizeProps.offset || sizeProps.offset === 0,
            [`${prefixCls}-${size}-push-${sizeProps.push}`]: sizeProps.push || sizeProps.push === 0,
            [`${prefixCls}-${size}-pull-${sizeProps.pull}`]: sizeProps.pull || sizeProps.pull === 0
        };
    });

    const classes = classNames(prefixCls, className, sizeClassObj, {
        [`${prefixCls}-${span}`]: span !== undefined,
        [`${prefixCls}-order-${order}`]: order,
        [`${prefixCls}-offset-${offset}`]: offset,
        [`${prefixCls}-push-${push}`]: push,
        [`${prefixCls}-pull-${pull}`]: pull
    });

    if (gutter) {
        style = {
            ...style,
            ...(gutter[0]! > 0
                ? { paddingLeft: gutter[0]! / 2, paddingRight: gutter[0]! / 2 }
                : {}),
            ...(gutter[1]! > 0 ? { paddingTop: gutter[1]! / 2, paddingBottom: gutter[1]! / 2 } : {})
        };
    }
    if (flex) {
        style.flex = parseFlex(flex);
    }

    return (
        <div className={classes} style={style} {...restProps}>
            {children}
        </div>
    );
};

Grid.Container = GridContainer;
Grid.displayName = "MR_Grid";

export default Grid;
