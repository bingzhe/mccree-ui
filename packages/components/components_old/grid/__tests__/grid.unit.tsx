import * as React from "react";
import { render, mount } from "enzyme";
import toJson from "enzyme-to-json";

import Grid from "../index";

describe("Grid", () => {
    it("should render Grid", () => {
        const wrapper = render(<Grid span={12} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it("should render Grid.Container", () => {
        const wrapper = render(<Grid.Container />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it("should work correct when gutter is object", () => {
        const wrapper = mount(<Grid.Container gutter={{ xs: 20 }} />);
        expect(wrapper.find("div").prop("style")).toEqual({
            marginLeft: -10,
            marginRight: -10
        });
    });
    it("should work currect when gutter is array", () => {
        const wrapper = mount(<Grid.Container gutter={[16, 20]} />);
        expect(wrapper.find("div").prop("style")).toEqual({
            marginLeft: -8,
            marginRight: -8,
            marginTop: -10,
            marginBottom: 10
        });
    });
});
