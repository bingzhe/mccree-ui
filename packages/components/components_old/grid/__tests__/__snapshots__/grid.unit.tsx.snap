// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid should render Grid 1`] = `
<div
  class="mccree-grid mccree-grid-12"
/>
`;

exports[`Grid should render Grid.Container 1`] = `
<div
  class="mccree-grid-container"
/>
`;
